agent/xds/failover_math.go

```diff
@@ -0,0 +1,34 @@
+package xds
+
+import (
+	envoycore "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+func firstHealthyTarget(
+	targetConfigs map[structs.DiscoveryTarget]structs.DiscoveryTargetConfig,
+	targetHealth map[structs.DiscoveryTarget]structs.CheckServiceNodes,
+	primary structs.DiscoveryTarget,
+	secondary []structs.DiscoveryTarget,
+) structs.DiscoveryTarget {
+	series := make([]structs.DiscoveryTarget, 0, len(secondary)+1)
+	series = append(series, primary)
+	series = append(series, secondary...)
+
+	for _, target := range series {
+		targetConfig := targetConfigs[target]
+
+		endpoints, ok := targetHealth[target]
+		if !ok {
+			continue
+		}
+		for _, ep := range endpoints {
+			healthStatus, _ := calculateEndpointHealthAndWeight(ep, targetConfig.Subset.OnlyPassing)
+			if healthStatus == envoycore.HealthStatus_HEALTHY {
+				return target
+			}
+		}
+	}
+
+	return primary // if everything is broken just use the primary for now
```

Comment:
What do you think about making the healthy threshold configurable? That was what we talked about in the original domain model RFC (i.e. failover if some percentage of nodes is unhealthy).

It seems simple to built here, I think the big downside is that Envoy doesn't exactly support that model so if it does fix weighted clusters to be able to set separate SNI per leg or similar we won't be able to use it as it won't exactly provide "healthy threshold" semantics.

Given that no other proxies work exactly like Envoy in this sliding failover case, what do you think about introducing a `healthy_threshold_percent` now which is the percentage of services that must be healthy to keep traffic flowing to them and not failover. Then if we want to re-introduce more complex "sliding" mechanism for Envoy later we could be explicit and add it as a separate mutually exclusive option like `sliding_failover_overprovision_factor`.

I think that's OK UX esp. given that this mode is one we can always implement for any proxy as we are in control and most other proxies are more likely to be able to do something like this than they are to do something with same semantics as Envoy over provisioning/weighting?

If the answer is that we'd rather ship like this I think that's OK but I feel like "wait til every single instance is dead" although that's what prepared queries do is not really an ideal strategy in a modern Service Mesh. It basically only works with extremely low numbers of replicas and low request volume. Everything else is just a prolonged cascading failure until at last the remaining nodes are overwhelmed and fail health checks and traffic shifts but then the overwhelmed nodes catch up and get "healthy" and traffic flaps back etc...

Reply:
Moving this to a new issue as it's purely additive: https://github.com/hashicorp/consul/issues/6276